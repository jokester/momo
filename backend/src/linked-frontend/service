../../submodules/momocraft-frontend/src/service