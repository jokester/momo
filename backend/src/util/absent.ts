export function absent(wanted: string): never {
  throw new Error(`expected ${wanted}`);
}
