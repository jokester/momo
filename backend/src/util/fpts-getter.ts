import { Either, isLeft } from 'fp-ts/lib/Either';
import { Option, isNone } from 'fp-ts/lib/Option';

export function getRightOrThrow<A, B>(e: Either<A, B>, onLeft: (l: A) => Error): B {
  if (isLeft(e)) {
    throw onLeft(e.left);
  }
  return e.right;
}

export function getSomeOrThrow<A>(o: Option<A>, onEmpty: () => Error): A {
  if (isNone(o)) {
    throw onEmpty();
  }
  return o.value;
}
