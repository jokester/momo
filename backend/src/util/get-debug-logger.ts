import path from 'path';
import debug from 'debug';
import { getDebugNamespace } from '@jokester/ts-commonutil/cjs/logging/get-debug-namespace';

const createNamespace = getDebugNamespace(path.join(__dirname, '..'));
export function getDebugLogger(srcFile: string): debug.IDebugger {
  const namespace = 'hanko:' + createNamespace(srcFile);
  return debug(namespace);
}
