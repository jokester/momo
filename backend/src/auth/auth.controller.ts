import { BadRequestException, Body, Controller, Get, Post, Req, Request, UseGuards } from '@nestjs/common';
import { AuthService } from './auth.service';
import { UserEntity } from '../entity/user.entity';
import { LocalStrategyGuard } from './local.strategy';
import { getDebugLogger } from '../util/get-debug-logger';
import { GoogleOAuthService } from './google-oauth.service';
import { UserService } from '../user/user.service';

const logger = getDebugLogger(__filename);

@Controller('auth')
export class AuthController {
  constructor(
    private readonly authService: AuthService,
    private readonly googleOAuthService: GoogleOAuthService,
    private readonly userService: UserService,
  ) {}

  @UseGuards(LocalStrategyGuard)
  @Post('email/signin')
  async emailSignIn(@Request() req: Request & { user?: UserEntity }): Promise<UserEntity | null> {
    logger('got email user: %o', req.user);
    return req.user || null;
  }

  @Get('google/oauth' /* ?code=callback_code... */)
  async startGoogleOAuth(@Request() req: Request & { user?: UserEntity }) {
    logger('got google/oauth user: %o', req.user);
    return null;
  }

  @Get('ping')
  async pingDb() {
    await this.userService.now();
    return null;
  }

  @Post('oauth/google')
  async doGoogleOAuth(@Request() req: Request, @Body() payload2: { code: string; callbackUri: string }) {
    // const payload: { code: string; callbackUri: string } = await req.json();

    const payload = payload2;
    if (payload && payload.code) {
      const x = await this.googleOAuthService.auth(payload.code);

      logger('got auth', x);
      return x;
    }
    throw new BadRequestException();
  }
}
