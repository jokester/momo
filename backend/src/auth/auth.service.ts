import { Injectable } from '@nestjs/common';
import { getDebugLogger } from '../util/get-debug-logger';

@Injectable()
export class AuthService {
  constructor() {
    getDebugLogger(__filename)('authService created');
  }

  touch() {
    getDebugLogger(__filename)('authService touched');
  }
}
