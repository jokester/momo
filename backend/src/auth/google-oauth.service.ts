import { google } from 'googleapis';
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { getDebugLogger } from '../util/get-debug-logger';
import { oauth2 } from 'googleapis/build/src/apis/oauth2';

const logger = getDebugLogger(__filename);

@Injectable()
export class GoogleOAuthService {
  constructor(private readonly confService: ConfigService) {}

  async auth(code: string) {
    const client = this.createAuthClient();
    logger('client=%O', client);
    logger('code=', code);

    const token = await client.getToken(code);
    client.setCredentials(token.tokens);
    logger('token=%O', token);

    const x = await oauth2({ version: 'v2', auth: client });
    const got = await x.userinfo.get();
    return got.data;
  }

  private createAuthClient() {
    return new google.auth.OAuth2({
      clientId: this.confService.get('OAUTH_GOOGLE_CLIENT_ID'),
      clientSecret: this.confService.get('OAUTH_GOOGLE_CLIENT_SECRET'),
      redirectUri: this.confService.get('OAUTH_GOOGLE_REDIRECT_URI'),
    });
  }
}
