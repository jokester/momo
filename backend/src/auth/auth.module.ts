import { Module } from '@nestjs/common';
import { AuthController } from './auth.controller';
import { DiscordOAuth, GoogleOAuth } from './oauth-client.provider';
import { DiscordOAuthService } from './discord-oauth.service';
import { GoogleOAuthService } from './google-oauth.service';
import { UserModule } from '../user/user.module';

@Module({
  imports: [UserModule],
  controllers: [AuthController],
  providers: [DiscordOAuth.Provider, DiscordOAuthService, GoogleOAuth.Provider, GoogleOAuthService],
})
export class AuthModule {}
