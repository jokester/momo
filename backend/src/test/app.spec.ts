import request from 'supertest';
import { Test } from '@nestjs/testing';
import { AppModule } from '../app.module';
import { INestApplication } from '@nestjs/common';
import { GoogleOAuthResponse, GoogleOAuthService } from '../user/google-oauth.service';
import { MockData, TestDeps } from './test-deps';
import { right } from 'fp-ts/lib/Either';
import { JwtService } from '@nestjs/jwt';
import { getDebugLogger } from '../util/get-debug-logger';
import { UserService } from '../user/user.service';
import { AuthController, AuthSuccessRes } from '../user/auth.controller';
import { MomoUserController } from '../momo/momo-user.controller';
import { getSomeOrThrow } from '../util/fpts-getter';
import { absent } from '../util/absent';
import { EmailAuthRequestDto } from '../model/auth.dto';
import { DiscordOAuth } from '../user/oauth-client.provider';
import { buildTesteeAppBundle } from './test-app-factory';

const logger = getDebugLogger(__filename);

describe('AppController (e2e)', () => {
  const { testBundle, bundledAfterAll, bundledBeforeAll } = buildTesteeAppBundle();
  let app: INestApplication;
  let jwtService: JwtService;
  let userService: UserService;

  beforeAll(async () => {
    await bundledBeforeAll();
    app = testBundle.app;
    jwtService = testBundle.jwtService;
    userService = testBundle.userService;
  });

  beforeEach(TestDeps.clearTestDatabase);

  afterAll(bundledAfterAll);

  describe(AuthController, () => {
    it('POST /auth/oauth/google returns jwtToken on succeed', async () => {
      jest
        .spyOn(GoogleOAuthService.prototype, 'auth')
        .mockResolvedValue(right<string, GoogleOAuthResponse>(MockData.googleOAuthResponseValid));

      const res = await request(app.getHttpServer()).post('/auth/oauth/google').send(MockData.oauthRequest).expect(201);

      const jwtToken = JSON.parse(res.text).jwtToken;
      expect(jwtToken).toBeTruthy();

      const decoded = await jwtService.decode(jwtToken);
      logger('decoded', decoded);

      await jwtService.verifyAsync(jwtToken);
    });

    it('POST /auth/oauth/google return 400 on malformed request', async () => {
      await request(app.getHttpServer()).post('/auth/oauth/google').send({ code: '123', redirectUrl: '' }).expect(400);
    });

    it('POST /auth/oauth/google return 400 on auth error', async () => {
      jest
        .spyOn(GoogleOAuthService.prototype, 'auth')
        .mockResolvedValue(right<string, GoogleOAuthResponse>(MockData.googleOAuthResponseEmailUnverified));

      const authErrorRes = await request(app.getHttpServer())
        .post('/auth/oauth/google')
        .send({ code: '123', redirectUrl: 'someUrl' })
        .expect(400);

      expect(authErrorRes.body).toMatchSnapshot('auth 400 res');
    });

    it('POST /auth/email/signup creates user and returns 201', async () => {
      {
        // signup success
        const created = await request(app.getHttpServer())
          .post('/auth/email/signup')
          .send(MockData.authPayload)
          .expect(201);

        const res: AuthSuccessRes = JSON.parse(created.text);
        expect(res.jwtToken).toBeTruthy();
        expect(res.user).toMatchSnapshot('POST /auth/email/signup 201');
      }

      {
        // signup fail: email must be unique
        await request(app.getHttpServer())
          .post('/auth/email/signup')
          .send({ email: 'a@B.com', password: '1234567' } as EmailAuthRequestDto)
          .expect(400);
      }
    });

    it('POST /auth/email/signin auths user', async () => {
      {
        /// signin success
        const signedIn = await request(app.getHttpServer())
          .post('/auth/email/signin')
          .send(MockData.authPayload)
          .expect(200);

        const res: AuthSuccessRes = JSON.parse(signedIn.text);
        expect(res.jwtToken).toBeTruthy();
        expect(res.user).toMatchSnapshot('POST /auth/email/signin 200');
      }

      {
        // signin fail: password incorrect
        await request(app.getHttpServer())
          .post('/auth/email/signin')
          .send({ ...MockData.authPayload, password: '123456789' })
          .expect(400);

        // signin fail: email incorrect
        await request(app.getHttpServer())
          .post('/auth/email/signin')
          .send({ ...MockData.authPayload, email: 'b@c.com' })
          .expect(400);
      }
    });
  });

  describe.skip(MomoUserController, () => {
    it('GET /user/:userId returns 404', async () => {
      await request(app.getHttpServer()).get('/user/__s').expect(404);
    });

    it('GET /user/self with improper auth header returns 400', async () => {
      await request(app.getHttpServer()).get('/user/self').set('Authorization', `Bear`).expect(400);
    });

    it('GET /user/self without auth return 401', async () => {
      await request(app.getHttpServer()).get('/user/self').expect(401);
    });

    it('GET /user/self with proper auth returns resolved user', async () => {
      const userAccount1 = getSomeOrThrow(await userService.findUser({ userId: '' }), () => absent('user by shortId'));

      const jwtToken = await userService.createJwtTokenForUser(userAccount1);

      await request(app.getHttpServer()).get('/user/self').set('Authorization', `Bearer ${jwtToken}`).expect(200);

      await request(app.getHttpServer()).get('/momo/user/self').set('Authorization', `Bearer ${jwtToken}`).expect(200);
    });

    it('PUT /user/self updated resolved user', async () => {
      const userAccount1 = getSomeOrThrow(await userService.findUser({ userId: '' }), () => absent('user by shortId'));

      const jwtToken = await userService.createJwtTokenForUser(userAccount1);

      const { body } = await request(app.getHttpServer())
        .put('/user/self')
        .set('Authorization', `Bearer ${jwtToken}`)
        .send({ nickName: 'samuel mf jackson', avatarUrl: 'https://corp.com/a.gif' })
        .expect(200);

      expect(body).toMatchSnapshot('updated user');

      const { body: body2 } = await request(app.getHttpServer())
        .get('/user/self')
        .set('Authorization', `Bearer ${jwtToken}`)
        .expect(200);

      expect(body2).toEqual(body);
    });
  });
});
