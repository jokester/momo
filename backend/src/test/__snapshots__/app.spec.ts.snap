// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController (e2e) AuthController POST /auth/email/signin auths user: POST /auth/email/signin 200 1`] = `
Object {
  "nickName": "(not set)",
  "shortId": "nanoid-2",
}
`;

exports[`AppController (e2e) AuthController POST /auth/email/signup creates user and returns 201: POST /auth/email/signup 201 1`] = `
Object {
  "nickName": "(not set)",
  "shortId": "nanoid-2",
}
`;

exports[`AppController (e2e) AuthController POST /auth/oauth/google returns jwtToken on succeed: GET /user/self 1`] = `
Object {
  "avatarUrl": "https://example.com/a.png",
  "nickName": "(not set)",
  "shortId": "nanoid-1",
}
`;

exports[`AppController (e2e) UserController PUT /user/self updated resolved user: updated user 1`] = `
Object {
  "avatarUrl": "https://corp.com/a.gif",
  "nickName": "samuel mf jackson",
  "shortId": "nanoid-1",
}
`;
